### PR TITLE
✨ Add Attach - Detach Options to Canvas Context Menu

### PR DESCRIPTION
**Pull request by MFA-X-AI**
# Description

This PR introduces 3 new context menu options, with the following spawn logic:
- **Attach** - will spawn if one or more `Literals` are selected.
- **Attach Literals** - will spawn if one or more `non-parameter components` are selected that has at least one detached Literal connected.
- **Detach Literals** -  will spawn if one or more `non-parameter components` are selected that has at least one attached Literal connected.

attach detach context options

## Pull Request Type

- [ ] Xircuits Core (Jupyterlab Related changes)
- [x] Xircuits Canvas (Custom RD Related changes)
- [ ] Xircuits Component Library
- [ ] Xircuits Project Template
- [ ] Testing Automation
- [ ] Documentation
- [ ] Others (Please Specify)

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Tests

1. Verify that you can `Attach` a Literal node by right clicking a Literal Node. Verify you can also do this by selecting multiple Literals.
2. Verify that you can attach Literals connected but detached from a component node. Verify that you can also do this by selecting multiple component nodes. 
3. Verify that you can detach Literals from a component node. Verify that you can also do this by selecting multiple component nodes. 

## Tested on?

- [x] Windows  
- [ ] Linux Ubuntu 
- [ ] Centos 
- [ ] Mac  
- [ ] Others  (State here -> xxx )

### CODE REVIEW

File: src/commands/NodeActionCommands.tsx [hunk 1]
```diff
@@ -910,4 +910,93 @@ export function addNodeActionCommands(
         
         }
     }
+
+    // Add command to attach selected node
+    commands.addCommand(commandIDs.attachNode, {
+        execute: async () => {
```
Comment (treo) [inline src/commands/NodeActionCommands.tsx:916]: I think this command should have a guard to not try and attach any nodes that aren't literals.

File: src/commands/NodeActionCommands.tsx [hunk 2]
```diff
@@ -910,4 +910,93 @@ export function addNodeActionCommands(
...
+
+            const widget = tracker.currentWidget?.content as XircuitsPanel;
+            const model = widget.xircuitsApp.getDiagramEngine().getModel();
+            const selected_entities = model.getSelectedEntities();
+            const selected_nodes = selected_entities.filter(entity => entity instanceof NodeModel) as CustomNodeModel[];
+            selected_nodes.forEach(node => {
+                node.getOptions().selected = false
```
Comment (treo) [inline src/commands/NodeActionCommands.tsx:923]: There is a `setSelected(val)` node, we might as well use it. I know it does the same thing under the hood, but still `node.setSelected(false)` reads better.

File: src/components/port/CustomPortModel.ts [hunk 3]
```diff
@@ -365,6 +366,26 @@ export  class CustomPortModel extends DefaultPortModel  {
 
     }
 
+    getTargetPorts = () => {
```
Comment (treo) [inline src/components/port/CustomPortModel.ts:369]: Neat, these functions could cut down on some of the repetition I've got in the `addHover` and `removeHover` functions, as well as the ones in the copy & paste support.

File: src/context-menu/CanvasContextMenu.tsx [hunk 4]
```diff
@@ -30,6 +31,21 @@ export class CanvasContextMenu extends React.Component<CanvasContextMenuProps> {
             });
         };
 
+        const handleAttachNode = async () => {
+            this.props.app.commands.execute(commandIDs.attachNode);
```
Comment (treo) [inline src/context-menu/CanvasContextMenu.tsx:35]: Is there a reason we don't await the first execution too?

File: src/context-menu/CanvasContextMenu.tsx [hunk 5]
```diff
@@ -93,7 +134,10 @@ export function getMenuOptionsVisibility(models) {
     let isSingleParameterNodeSelected = parameterNodes.length === 1;
     let isSingleComponentNodeSelected = componentNodes.length === 1;
     let showReloadNode = isNodeSelected && componentNodes.length > 0;
-    let showopenXircuitsWorkflow = isSingleComponentNodeSelected && models.some(model => isXircuitsWorkflow(model))
+    let showopenXircuitsWorkflow = isSingleComponentNodeSelected && models.some(model => isXircuitsWorkflow(model));
+    let showAttachNode = isNodeSelected && parameterNodes.length > 0;
```
Comment (treo) [inline src/context-menu/CanvasContextMenu.tsx:138]: Do I understand this correctly that this will show up on non literal nodes too?

**Review by treo — CHANGES_REQUESTED** [review]

File: src/commands/NodeActionCommands.tsx (hunk 1 above)
Comment (MFA-X-AI) [reply src/commands/NodeActionCommands.tsx:916]: Updated so only entities that starts with "Literal  " is considered.

File: src/commands/NodeActionCommands.tsx (hunk 2 above)
Comment (MFA-X-AI) [reply src/commands/NodeActionCommands.tsx:923]: Ah good one. Updated. Maybe we'd want a set attached method in the future.

File: src/context-menu/CanvasContextMenu.tsx (hunk 4 above)
Comment (MFA-X-AI) [reply src/context-menu/CanvasContextMenu.tsx:35]: Good question. The attach / detach is quite light, but to be safe we should await it.
Updated.

File: src/context-menu/CanvasContextMenu.tsx (hunk 5 above)
Comment (MFA-X-AI) [reply src/context-menu/CanvasContextMenu.tsx]: Ah good one. Currently it shows when at least one parameter node is selected, but it should be when at least one literal node is selected as arguments can't be attached. 
One thing I missed was the context being able to banish Literals to the shadow realm if they're not connected. That logic has been fixed.

File: src/components/port/CustomPortModel.ts (hunk 3 above)
Comment (MFA-X-AI) [reply src/components/port/CustomPortModel.ts:369]: For this one, I've noticed that the `selectAllRelevantNodes` also highlights the links, and the addHover - removeHover accesses the properties through their names.

**Review by treo — APPROVED** [review]

**Pull request merged**